### ISSUE TITLE
[Bug] create extension gp_debug_numsegments changes current search_path

### ISSUE THREAD

**Issue by tuhaihe** [cloudberrydb/cloudberrydb#59]
### Cloudberry Database version

Greenplum version 7

### What happened

when creating extension `gp_debug_numsegments` changes current search_path, drop currrent search_path.

### What you think should happen instead

Should not drop current search_path.

### How to reproduce

```sql
gpadmin=# show search_path ;
   search_path
-----------------
 "$user", public
(1 row)

gpadmin=# create schema xmen;
CREATE SCHEMA
gpadmin=# set search_path to xmen;
SET
gpadmin=# show search_path ;
 search_path
-------------
 xmen
(1 row)

gpadmin=# CREATE EXTENSION IF NOT EXISTS gp_debug_numsegments;
CREATE EXTENSION
gpadmin=# show search_path ;
 search_path
-------------
 public
(1 row)
```

See https://github.com/greenplum-db/gpdb/issues/15912

### Operating System

Linux

### Are you willing to submit PR?

- [ ] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://cloudberrydb.org/community/coc).

**Comment by Ray-Eldath** [comment]
This issue seems have nothing to do with the actual extension being installed. Using `gp_sparse_vector` can lead to the same issue:
<img width="419" alt="image" src="https://github.com/cloudberrydb/cloudberrydb/assets/7938763/bf836ea5-d715-4608-a702-4e7a8791c9ff">
My guess here is that a extension will cause this **if** they are relocatable (`relocatable = true`) in their `.control` file but using `SET search_path = xxx` in their script (`<name>--<version>.sql` file). Changes:

```diff
Index: src/backend/commands/extension.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/backend/commands/extension.c b/src/backend/commands/extension.c
--- a/src/backend/commands/extension.c	(revision c9dbca2e5beaaf9854abaebcfd600aa77780f40b)
+++ b/src/backend/commands/extension.c	(date 1690454308496)
@@ -1121,7 +1121,7 @@
 		/*
 		 * Restore the GUC variables we set above.
 		 */
-		AtEOXact_GUC(true, save_nestlevel);
+		AtEOXact_GUC(false, save_nestlevel);
 	}
 	PG_END_TRY();
```

can surely fix the issue by pretending it's at a transaction abort (instead of a commit) to force prior values restoration, but changes to Postgres kernel are unwanted. It's still unknown whether a extension: _i._  states itself "relocatable" in control file; _ii._ set `search_path` w/o using `@extschema@` is _truly_ "relocatable" in any sensible manner.

**Comment by Ray-Eldath** [comment]
A normal extension who doesn't use `SET search_path` in its script file should install the objects to the environment `search_path` in which the `CREATE EXTENSION` were executed:
<img width="1355" alt="image" src="https://github.com/cloudberrydb/cloudberrydb/assets/7938763/12c9951f-72a6-49f2-9642-8b82b1f573d9">


But these "malformed" extension are installing objects to wherever they want, regardless of the `search_path` of the env:
<img width="1011" alt="image" src="https://github.com/cloudberrydb/cloudberrydb/assets/7938763/0fc860f8-cb17-4731-ac6e-3b0d418e39b8">


They're also unresponsive to the explicitly-given `SCHEMA` option:
<img width="1007" alt="image" src="https://github.com/cloudberrydb/cloudberrydb/assets/7938763/67ea43bc-5ba5-4894-b482-75964d9e69de">

It could be argued that these extensions are **not-well-written** in the first place.

**Comment by avamingli** [comment]
> It could be argued that these extensions are not well written in the first place.

+1, and have a try with other PG/GP 's extension to verify?

**Comment by Ray-Eldath** [comment]
YES I've tried.  both `gp_debug_numsegments` and `gp_sparse_vector` show the same "problematic" behavior:
<img width="440" alt="image" src="https://github.com/cloudberrydb/cloudberrydb/assets/7938763/7330e9cc-6a77-4d8d-9442-d0bf9460e6ec">
These two extensions use `SET search_path = xxx` w/o restoring them. It's actually [**recommended as the best practice** to temporarily set `search_path` to `pg_catalog, pg_temp` to guard against "search-path-based attacks"](https://www.postgresql.org/docs/current/extend-extensions.html#id-38.17.6.1), and many extensions indeed do the trick:
<img width="1461" alt="image" src="https://github.com/cloudberrydb/cloudberrydb/assets/7938763/43ff765d-e9dc-4be1-ae58-3e8acbba9b97">
I suggest the proposed fix to be just adding a "restoration to prior `search_path`" SQL statement to the problematic extensions, just as all these "normal" extensions.

**Comment by avamingli** [comment]
Do pg extensions change user's search_path?

**Comment by Ray-Eldath** [comment]
> Do pg extensions change user's search_path?
 
They probably shouldn't 🤔 .

**Comment by Ray-Eldath** [comment]
> Do pg extensions change user's search_path?

I've write a small script to test all extensions listed in `pg_available_extensions`, the following is the result:
<img width="970" alt="image" src="https://github.com/cloudberrydb/cloudberrydb/assets/7938763/979dff61-cc6b-48ad-b634-50642cff182c">
```
pgrowlocks Pass
gp_sparse_vector Failed
fuzzystrmatch Pass
btree_gist Pass
gp_distribution_policy Failed
intagg Pass
btree_gin Pass
hstore_plpython3u Pass
gp_exttable_fdw Pass
unaccent Pass
tsm_system_time Pass
tcn Pass
citext Pass
hstore_plpython2u ERROR
dict_xsyn Pass
pxf_fdw Pass
tsm_system_rows Pass
pg_stat_statements Pass
bloom Pass
pg_trgm Pass
pg_freespacemap Pass
seg ERROR
hstore Duplicated
pg_prewarm Pass
gp_debug_numsegments Failed
insert_username Pass
isn Pass
pg_surgery Pass
gp_internal_tools Failed
gp_replica_check Pass
intarray Pass
pgstattuple Pass
moddatetime Pass
amcheck ERROR
ltree_plpythonu ERROR
jsonb_plpython2u ERROR
gp_inject_fault ERROR
gp_legacy_string_agg Pass
hstore_plpythonu ERROR
plpgsql Pass
plpython3u Duplicated
pgcrypto Pass
ltree_plpython3u Pass
tablefunc Pass
pg_visibility Pass
autoinc Pass
ltree_plpython2u ERROR
pageinspect Pass
ltree Duplicated
old_snapshot Pass
pg_buffercache Pass
dblink Pass
jsonb_plpython3u Pass
file_fdw Pass
refint Pass
dict_int Pass
jsonb_plpythonu ERROR
```
seems `gp_sparse_vector`, `gp_distribution_policy`, `gp_debug_numsegments` and `gp_internal_tools` has failed on the test. Notably, GP extensions `gp_exttable_fdw`, `gp_replica_check` and `gp_legacy_string_agg` do pass the test.

So the answer is YES. all extensions that are **not** started with the `gp_` prefix preserves the user's `search_path` whereas **some of (not all of)** GP extensions do not.
